Mage.Sets/src/mage/cards/c/CovetedFalcon.java

```diff
@@ -0,0 +1,152 @@
+        super(ownerId, setInfo, new CardType[]{CardType.ARTIFACT, CardType.CREATURE}, "{1}{U}{U}");
+
+        this.subtype.add(SubType.BIRD);
+        this.power = new MageInt(1);
+        this.toughness = new MageInt(4);
+
+        // Flying
+        this.addAbility(FlyingAbility.getInstance());
+
+        // Whenever Coveted Falcon attacks, gain control of target permanent you own but don't control.
+        Ability attacksTriggeredAbility = new AttacksTriggeredAbility(new GainControlTargetEffect(Duration.Custom));
+        attacksTriggeredAbility.addTarget(new TargetPermanent(filter));
+        this.addAbility(attacksTriggeredAbility);
+
+        // Disguise {1}{U}
+        this.addAbility(new DisguiseAbility(this, new ManaCostsImpl<>("{1}{U}")));
+
+        // When Coveted Falcon is turned face up, target opponent gains control of any number of target permanents you control.
+        // Draw a card for each one they gained control of this way.
+        Ability turnedFaceUpTriggeredAbility = new TurnedFaceUpSourceTriggeredAbility(new CovetedFalconEffect());
+        turnedFaceUpTriggeredAbility.addTarget(new TargetOpponent());
+        turnedFaceUpTriggeredAbility.addTarget(
+                new TargetControlledPermanent(0, Integer.MAX_VALUE, StaticFilters.FILTER_CONTROLLED_PERMANENT, false));
+        this.addAbility(turnedFaceUpTriggeredAbility);
+    }
+
+    private CovetedFalcon(final CovetedFalcon card) {
+        super(card);
+    }
+
+    @Override
+    public CovetedFalcon copy() {
+        return new CovetedFalcon(this);
+    }
+}
+
+class CovetedFalconEffect extends ContinuousEffectImpl {
+
+    private boolean firstControlChange = true;
+    private List<UUID> permanentIds = new ArrayList<>();
+
+    CovetedFalconEffect() {
+        super(Duration.Custom, Layer.ControlChangingEffects_2, SubLayer.NA, Outcome.Benefit);
+        staticText = "target opponent gains control of any number of target permanents you control. " +
+                "Draw a card for each one they gained control of this way";
+    }
+
+    private CovetedFalconEffect(final CovetedFalconEffect effect) {
+        super(effect);
+        firstControlChange = effect.firstControlChange;
+        permanentIds = effect.permanentIds;
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            discard(); // controller no longer exists
+            return false;
+        }
+
+        UUID controllingOpponentId = source.getFirstTarget();
+        if (firstControlChange) {
+            permanentIds.addAll(source.getTargets().get(1).getTargets());
+        }
+
+        int cardsToDraw = 0;
+        ListIterator<UUID> permanentsIterator = permanentIds.listIterator(permanentIds.size());
+        boolean oneTargetStillExists = false;
+        while (permanentsIterator.hasPrevious()) {
```

Comment:
I'm not familiar with this style, can you explain why `hasPrevious()` is the preferred approach here?

Reply:
I traversed the list backwards because I didn't want to accidentally skip elements in the list when the current one has just been removed.